refactor(core): add input and output filtering for host directives

crisbeto opened the pull request:
Adds the logic that will filter out unexposed inputs/outputs and apply the aliases that the author specified when writing the host directives.

crisbeto commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1265,13 +1290,14 @@ export function invokeHostBindingsInCreationMode(def: DirectiveDef<any>, directi
-    tView: TView, lView: LView,
-    tNode: TElementNode|TContainerNode|TElementContainerNode): DirectiveDef<unknown>[]|null {
+    tView: TView, lView: LView, tNode: TElementNode|TContainerNode|TElementContainerNode):
+    [matches: DirectiveDef<unknown>[], definitionMap: HostDirectiveDefinitionMap|null]|null {
```
I'm not a fan of having this return an array when it might not need to, but I couldn't find a better way of producing two results from `findDirectiveDefMatches`. Some of the considered alternatives:
* Having a separate function that is called inside `resolveDirectives` that produces the map. The problem with this is that we don't have a way of distinguishing which matches came from host directives versus selector matching.
* Inverting the selector matching logic. We decided against doing so here https://github.com/angular/angular/pull/47228#pullrequestreview-1082113174.
* Have `findDirectiveDefMatches` return something like `DirectiveDef[] | {matches: DirectiveDef[], mappings: HostDirectiveDefinitionMap} | null`. This should be viable, but I'm not sure about the perf implications of generating object literals instead of arrays.

AndrewKushnir replied:
Just as an idea, we can pass an object as an accumulator into the `findDirectiveDefMatches` function, for example:
```
const accumulator = {matchedDefs: null, hostDirectiveDefs: null};
collectDirectiveDefMatches(accumulator, ...);
```

I think the current approach is fine too, we do have the `assertFirstCreatePass` at the very beginning of the function, so we guarantee to invoke the function only once per component (at creation time), so it should not affect the performance.

crisbeto replied:
That was one of the approaches I was considering, but I decided against it because we'd be creating the object even if we don't need it. That being said, I feel like I've been microoptimizing a bit here.

crisbeto commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -412,12 +413,27 @@ export interface HostDirectiveDef<T = unknown> {
+ * Mapping between a directive that was used as a host directive
+ * and the configuration that was used to define it as such.
+ */
+export type HostDirectiveDefinitionMap = Map<DirectiveDef<unknown>, HostDirectiveDef>;
```
I was struggling a bit with how to name this so I'm open to suggestions.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -863,28 +863,39 @@ export function createTNode(
-        propStore[publicName].push(directiveDefIdx, internalName);
-      } else {
-        (propStore[publicName] = [directiveDefIdx, internalName]);
+      if (aliasMap === null) {
+        addToPropStore(propStore, directiveIndex, publicName, internalName);
+      } else if (aliasMap.hasOwnProperty(publicName)) {
+        addToPropStore(propStore, directiveIndex, aliasMap[publicName], internalName);
       }
```
Proposal to improve readability (and a request to add some docs):
```suggestion
      if (aliasMap !== null && aliasMap.hasOwnProperty(publicName)) {
        // ... docs on why we do it ...
        publicName = aliasMap[publicName];
      }
      addToPropStore(propStore, directiveIndex, publicName, internalName);
```

crisbeto replied:
This doesn't cover the same cases though. The host directive alias map has two purposes: it acts as an allowlist for the inputs/outputs and it remaps the bindings to a different name. In your example the allowlist isn't covered. I've added a comment explaining why it's written this way.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -863,28 +863,39 @@ export function createTNode(
-    propStore: PropertyAliases|null): PropertyAliases|null {
+    inputAliasMap: {[publicName: string]: string}, directiveIndex: number,
+    propStore: PropertyAliases|null,
+    aliasMap: {[internalName: string]: string}|null): PropertyAliases|null {
```
nit: I think it'd be great to have a more descriptive name here:
```suggestion
    hostDirectiveInputAliases: {[internalName: string]: string}|null): PropertyAliases|null {
```

AndrewKushnir replied:
This function would also benefit from having a JSDocs that explains the processing that happens inside.

crisbeto replied:
I added a JSDoc and some more explanation in the comments, but I don't agree about the new name since the mapping is used both for inputs and outputs.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -863,28 +863,39 @@ export function createTNode(
+  if (propStore.hasOwnProperty(publicName)) {
+    propStore[publicName].push(directiveIndex, internalName);
+  } else {
+    (propStore[publicName] = [directiveIndex, internalName]);
```
```suggestion
    propStore[publicName] = [directiveIndex, internalName];
```

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -898,8 +909,13 @@ export function initializeInputAndOutputAliases(tView: TView, tNode: TNode): voi
 
   for (let i = start; i < end; i++) {
```
Unrelated to the change in this PR (we can do it later): it'd be cool to rename `i` to `directiveIndex` :)

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1265,13 +1290,14 @@ export function invokeHostBindingsInCreationMode(def: DirectiveDef<any>, directi
 
   const registry = tView.directiveRegistry;
   let matches: DirectiveDef<unknown>[]|null = null;
+  let definitionMap: HostDirectiveDefinitionMap|null = null;
```
I'd propose renaming the `let` to be more specific:
```suggestion
  let hostDirectiveDefs: HostDirectiveDefs|null = null;
```

crisbeto commented on packages/core/src/render3/features/host_directives_feature.ts:
```diff
@@ -58,7 +58,8 @@ export function ɵɵHostDirectivesFeature(rawHostDirectives: HostDirectiveConfig
 
 function findHostDirectiveDefs(
-    matches: DirectiveDef<unknown>[], def: DirectiveDef<unknown>, tView: TView, lView: LView,
+    matches: DirectiveDef<unknown>[], definitionMap: HostDirectiveDefinitionMap,
+    def: DirectiveDef<unknown>, tView: TView, lView: LView,
```
I've renamed and reshuffled the arguments as proposed. I also realized that `tView`, `lView` and `tNode` aren't used anymore so I removed them. As for the JSDoc, there already is one in `packages/core/src/render3/interfaces/definition.ts`. I put it there since that's what'll show up in the autocompletion when you try to call the function. I've expanded the doc a bit.

AndrewKushnir replied:
Thanks for the reply. I think it'd still be great to add a note as a JSDoc to mention that we mutate the `matchedDefs` and `hostDirectiveDefs` within the function, so that it shows up in IDE as a hint. Not super important though, it'd be ok to skip too.

crisbeto commented:
The feedback has been addressed.

pkozlowski-opensource commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -861,30 +861,55 @@ export function createTNode(
   return propStore;
 }
 
+function addToPropStore(
```
nit: I'm finding the "PropStore" name super-confusing, especially when it is used alongside the `PropertyAliases`. How about we use the `PropertyAlias` name exclusively? So this function could be named `addPropertyAlias`?
(this could be done in a follow-up PR and I'm happy to do this - just trying to see if we can reduce the number of names / concepts that we are using...)

crisbeto replied:
Done.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -412,12 +417,27 @@ export interface HostDirectiveDef<T = unknown> {
 
   /** Directive inputs that have been exposed. */
-  inputs: {[publicName: string]: string};
+  inputs: HostDirectiveBindingMap;
```
If I'm not mistaken, the `inputs` / `outputs` field is applicable to directives that don't have any host directives applied to them, right? If so, I find the name `HostDirectiveBindingMap` misleading here. It is more about input / output aliasing, right?

crisbeto replied:
It does two things:
1. It acts as an allowlist for the inputs/outputs that are exposed.
2. It's used to remap the public name.

I was struggling with how to capture this in the name of the symbol so I went with something more generic like "binding".

crisbeto committed:
refactor(core): add input and output filtering for host directives

Adds the logic that will filter out unexposed inputs/outputs and apply the aliases that the author specified when writing the host directives.

crisbeto commented:
I ended up squashing the commits so it's easier to rebase the changes from #47530.

pkozlowski-opensource reviewed (approved):
LGTM overall.

Personally I had a different design on my mind where the feature is patching `setInput` and more of the code goes into the feature itself. But this would mean substantial re-design / re-write.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: size-tracking
Reviewed-for: fw-core

AndrewKushnir commented:
This PR was merged into the repository by commit 76a8c68cc1a924091b45112ad4f649fd264d9482.
